File: lib/release.js
import process from "node:process";
import semver from "semver";
import resources from "./resources.js";

/**
 * Represents a GitHub release.
 */
export class Release {

	/**
	 * The base URL of the releases.
	 * @type {URL}
	 */
	static #baseUrl = new URL("https://github.com/HaxeFoundation/hashlink/");

	/**
	 * The list of all releases.
	 * @type {Release[]}
	 */
	static #data = resources.map(item => new Release(item));

	/**
	 * The associated assets.
	 * @type {ReleaseAsset[]}
	 */
	assets;

	/**
	 * The version number.
	 * @type {string}
	 */
	version;

	/**
	 * Creates a new release.
	 * @param {ReleaseOptions} [options] An object providing values to initialize this instance.
	 */
	constructor(options = {}) {
		this.assets = options.assets ?? [];
		this.version = options.version ?? "0.0.0";
	}

	/**
	 * The latest release.
	 * @type {Release}
	 */
	static get latest() {
		return this.#data[0];
	}

	/**
	 * Value indicating whether this release exists.
	 * @type {boolean}
	 */
	get exists() {
		return Release.#data.some(release => release.version == this.version);
	}

	/**
	 * Value indicating whether this release is provided as source code.
	 * @type {boolean}
	 */
	get isSource() {
		return !this.getAsset(process.platform);
	}

	/**
	 * The associated Git tag.
	 * @type {string}
	 */
	get tag() {
		const [major, minor, patch] = this.version.split(".");
		return Number.parseInt(patch) > 0 ? `${major}.${minor}.${patch}` : `${major}.${minor}`;
	}

	/**
	 * The download URL.
	 * @type {URL}
	 */
	get url() {
		const asset = this.getAsset(process.platform);
		return new URL(asset ? `releases/download/${this.tag}/${asset.file}` : `archive/refs/tags/${this.tag}.zip`, Release.#baseUrl);
	}

	/**
	 * Finds a release that matches the specified version constraint.
	 * @param {string} constraint The version constraint.
	 * @returns {Release|null} The release corresponding to the specified version constraint.
	 */
	static find(constraint) {
		return this.#data.find(release => semver.satisfies(release.version, constraint)) ?? null;
	}

	/**
	 * Gets the release corresponding to the specified version.
	 * @param {string} version The version number of a release.
	 * @returns {Release|null} The release corresponding to the specified version.
	 */
	static get(version) {
		return this.#data.find(release => release.version == version) ?? null;
	}

	/**
	 * Gets the asset corresponding to the specified platform.
	 * @param {import("./platform.js").Platform} platform The target platform.
	 * @returns {ReleaseAsset|null} The asset corresponding to the specified platform.
	 */
	getAsset(platform) {
		return this.assets.find(asset => asset.platform == platform) ?? null;
	}
}

/**
 * Represents an asset of a GitHub release.
 * @typedef {object} ReleaseAsset
 * @property {string} file The target file.
 * @property {import("./platform.js").Platform} platform The target platform.
 */

/**
 * Defines the options of a {@link Release} instance.
 * @typedef {object} ReleaseOptions
 * @property {ReleaseAsset[]} [assets] The associated assets.
 * @property {string} [version] The version number.
 */

File: src/Release.cs
namespace Belin.SetupHashLink;

using System.Linq;
using System.Text.RegularExpressions;

/// <summary>
/// Represents a HashLink release.
/// </summary>
/// <param name="version">The version number.</param>
/// <param name="assets">The associated assets.</param>
public partial class Release(Version version, IEnumerable<Release.Asset>? assets = null): IEquatable<Release> {

	/// <summary>
	/// The latest release.
	/// </summary>
	public static Release Latest => data.First();

	/// <summary>
	/// Gets the regular expression used to check if a version number represents the latest release.
	/// </summary>
	/// <returns>The regular expression used to check if a version number represents the latest release.</returns>
	[GeneratedRegex(@"^(\*|latest)$", RegexOptions.IgnoreCase)]
	internal static partial Regex LatestReleasePattern();

	/// <summary>
	/// The associated assets.
	/// </summary>
	public IEnumerable<Asset> Assets => assets ?? [];

	/// <summary>
	/// Value indicating whether this release exists.
	/// </summary>
	public bool Exists => data.Any(release => release == this);

	/// <summary>
	/// Value indicating whether this release is provided as source code.
	/// </summary>
	public bool IsSource => GetAsset(PlatformExtensions.Current) is null;

	/// <summary>
	/// The associated Git tag.
	/// </summary>
	public string Tag => Version.ToString(Version.Build > 0 ? 3 : 2);

	/// <summary>
	/// The download URL.
	/// </summary>
	public Uri Url {
		get {
			var asset = GetAsset(PlatformExtensions.Current);
			var baseUrl = new Uri("https://github.com/HaxeFoundation/hashlink/");
			return new(baseUrl, asset is null ? $"archive/refs/tags/{Tag}.zip" : $"releases/download/{Tag}/{asset.File}");
		}
	}

	/// <summary>
	/// The version number.
	/// </summary>
	public Version Version => version;

	/// <summary>
	/// Creates a new release.
	/// </summary>
	/// <param name="version">The version number.</param>
	/// <param name="assets">The associated assets.</param>
	public Release(string version, IEnumerable<Asset>? assets = null): this(Version.Parse(version), assets) {}

	/// <summary>
	/// Determines whether the two specified objects are equal.
	/// </summary>
	/// <param name="object1">The first object.</param>
	/// <param name="object2">The second object.</param>
	/// <returns><see langword="true"/> if <c>object1</c> equals <c>object2</c>, otherwise <see langword="false"/>.</returns>
	public static bool operator ==(Release? object1, Release? object2) =>
		object1 is null ? object2 is null : ReferenceEquals(object1, object2) || object1.Equals(object2);

	/// <summary>
	/// Determines whether the two specified objects are not equal.
	/// </summary>
	/// <param name="object1">The first object.</param>
	/// <param name="object2">The second object.</param>
	/// <returns><see langword="true"/> if <c>object1</c> does not equal <c>object2</c>, otherwise <see langword="false"/>.</returns>
	public static bool operator !=(Release? object1, Release? object2) => !(object1 == object2);

	/// <summary>
	/// Finds a release that matches the specified version constraint.
	/// </summary>
	/// <param name="constraint">The version constraint.</param>
	/// <returns>The release corresponding to the specified constraint, or <see langword="null"/> if not found.</returns>
	/// <exception cref="FormatException">The version constraint is invalid.</exception>
	public static Release? Find(string constraint) {
		var operatorMatch = Regex.Match(constraint, @"^([^\d]+)\d");
		var (op, version) = true switch {
			true when LatestReleasePattern().IsMatch(constraint) => ("=", Latest.Version.ToString()),
			true when operatorMatch.Success => (operatorMatch.Groups[1].Value, Regex.Replace(constraint, @"^[^\d]+", "")),
			true when Regex.IsMatch(constraint, @"^\d") => (">=", constraint),
			_ => throw new FormatException("The version constraint is invalid.")
		};

		var semver = SemanticVersion.Parse(version);
		return data.FirstOrDefault(op switch {
			">" => release => new SemanticVersion(release.Version) > semver,
			">=" => release => new SemanticVersion(release.Version) >= semver,
			"=" => release => new SemanticVersion(release.Version) == semver,
			"<=" => release => new SemanticVersion(release.Version) <= semver,
			"<" => release => new SemanticVersion(release.Version) < semver,
			_ => throw new FormatException("The version constraint is invalid.")
		});
	}

	/// <summary>
	/// Gets the release corresponding to the specified version.
	/// </summary>
	/// <param name="version">The version number of a release.</param>
	/// <returns>The release corresponding to the specified version, or <see langword="null"/> if not found.</returns>
	public static Release? Get(string version) => Get(Version.Parse(version));

	/// <summary>
	/// Gets the release corresponding to the specified version.
	/// </summary>
	/// <param name="version">The version number of a release.</param>
	/// <returns>The release corresponding to the specified version, or <see langword="null"/> if not found.</returns>
	public static Release? Get(Version version) => data.SingleOrDefault(release => release.Version == version);

	/// <summary>
	/// Determines whether the specified object is equal to this object.
	/// </summary>
	/// <param name="other">An object to compare with this object.</param>
	/// <returns><see langword="true"/> if the specified object is equal to this object, otherwise <see langword="false"/>.</returns>
	public override bool Equals(object? other) => Equals(other as Release);

	/// <summary>
	/// Determines whether the specified object is equal to this object.
	/// </summary>
	/// <param name="other">An object to compare with this object.</param>
	/// <returns><see langword="true"/> if the specified object is equal to this object, otherwise <see langword="false"/>.</returns>
	public bool Equals(Release? other) => other is not null && Version == other.Version;

	/// <summary>
	/// Gets the asset corresponding to the specified platform.
	/// </summary>
	/// <param name="platform">The target platform.</param>
	/// <returns>The asset corresponding to the specified platform, or <see langword="null"/> if not found.</returns>
	public Asset? GetAsset(Platform platform) => Assets.SingleOrDefault(asset => asset.Platform == platform);

	/// <summary>
	/// Gets the hash code for this object.
	/// </summary>
	/// <returns>The hash code for this object.</returns>
	public override int GetHashCode() => HashCode.Combine(Version);
}
